Merge HeliAttack into FlyAttack

reaperrr opened the pull request:
Split from #15803.

Depends on #16369.

tovl commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -118,8 +132,33 @@ public override Activity Tick(Actor self)
-				QueueChild(self, new Fly(self, target, WDist.Zero, lastVisibleMaximumRange, checkTarget.CenterPosition, Color.Red), true);
+				// Fly towards the desired facing (if CanHover) or last known position
+				if (aircraft.Info.CanHover)
+					aircraft.SetPosition(self, aircraft.CenterPosition + aircraft.FlyStep(desiredFacing));
```
Maybe this should queue `HeliFly` as a `ChildActivity` instead of doing all the positioning stuff itself?

reaperrr replied:
The problem with that, at least in this particular activity, is that it needs to check every tick whether it's close enough to the target now, so using a ChildActivity for movement would complicate things.

I agree that this activity shouldn't call SetPosition and TickFacing directly, though, and should at least use `Fly.FlyToward` from #16483 instead.

tovl replied:
But `HeliFly` already takes care of that, doesn't it? Why can't we just call `move.MoveWithinRange`, just like `AttackFollow` does?

reaperrr replied:
Thanks for pointing me in the right direction, adressed in #16483.

tovl commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -118,8 +132,33 @@ public override Activity Tick(Actor self)
+					else if (aircraft.Facing != 0)
+						facing = aircraft.Facing;
+					aircraft.SetPosition(self, aircraft.CenterPosition + aircraft.FlyStep(-facing));
+				}
```
Maybe this complete branch should queue `HeliFly` as a `ChildActivity` instead of doing all the positioning stuff itself?

reaperrr commented:
Waiting for #16483 to be merged first would likely save me from some serious rebase conflicts due to how HeliAttack currently handles movement, so I see little point in rebasing & updating this until #16483 is in.

reaperrr committed:
Merge HeliAttack into FlyAttack

And polish CanHover FlyAttack behavior:

- Get rid of direct TickFacing usage
- Fix that the CanHover facing/altitude update would override
  TakeOff child of Fly
- Streamline the queueing of child activities
- Get rid of a direct FlyTick in favor of relying on Fly activity
- Pull queueing of TakeOff out of the if-else

reaperrr commented:
Updated and rebased.

Note: In light of the experience (or should I say trauma...?) with the HeliFly merge, *please* lets stick to merging and - apart from the 2nd commit, which is limited in scope and therefore relatively easy to handle - fixing regressions, if there are any.

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -150,14 +150,39 @@ public override Activity Tick(Actor self)
+				QueueChild(self, new Fly(self, target, WDist.Zero, lastVisibleMaximumRange, checkTarget.CenterPosition, Color.Red), true);
+			else if (target.IsInRange(pos, attackAircraft.GetMinimumRangeVersusTarget(target)))
+			{
+				// Facing 0 doesn't work with the following position change
```
What is this code trying to do? Remember that facing 0 and 1 are real facings, -1 is used to mean "any facing".

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -150,14 +150,39 @@ public override Activity Tick(Actor self)
+				QueueChild(self, new FlyTimed(ticksUntilTurn, self));
+			}
+			else if (!isAirborne)
+				QueueChild(self, new TakeOff(self), true);
```
If !CanHover we queue the takeoff and the activities to do after takeoff in one go, whereas CanHover queus just the takeoff and waits until this next ticks to queue the movement.

We should be able to reduce duplication by putting
```csharp
if (!isAirborne)
	QueueChild(self, new TakeOff(self), true);
```

before the `if (!aircraft.Info.CanHover)` and letting both cases queue their movements after it?

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -150,14 +150,39 @@ public override Activity Tick(Actor self)
+				var move = aircraft.FlyStep(-facing);
+				Fly.FlyTick(self, aircraft, desiredFacing, aircraft.Info.CruiseAltitude, move);
+			}
+			else // CanHover and within firing range, so only update facing and altitude (if necessary)
```
It would improve readability if we format the code as
```csharp
if (!aircraft.Info.CanHover)
{
	if (...)
		...
	...
}
else
{
	if (...)
		...
	...
}
```
instead of having the different CanHover cases handled as `else if`s. This comment can probably then be removed as being obvious.

reaperrr commented:
Updated. I had the new commit almost ready anyway (though originally for a follow-up...).

If this regresses anything major/hard-to-fix, I'd rather drop it and defer those changes back to that follow-up again, though.

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -150,39 +150,29 @@ public override Activity Tick(Actor self)
 			}
 
-			var delta = attackAircraft.GetTargetPosition(pos, checkTarget) - pos;
+			var delta = attackAircraft.GetTargetPosition(pos, target) - pos;
```
Was replacing `checkTarget` with `target` intentional? This will regress targeting vs units that have disappeared under the fog (`target` refers to the *actual* hidden unit, `checkTarget` to its last known position).

reaperrr replied:
Yes, it actually was intentional.

If `useLastVisibleTarget` is `true`, we won't get here anyway (due to the `if` block above), and if it's false, `checkTarget` is equal to `target` anyway.
This was just about consistency (this and some places below were using checkTarget, while other parts below already used target).
I can change everything to `checkTarget` if you think that's better, but like I said, if we got this far in the activity they're equal anyway.

pchote replied:
Ah, now I see. What you have here is fine.

pchote reviewed (dismissed):
LGTM, but IMO it would be be better if the commits were squashed before merging.

reaperrr commented:
> but IMO it would be be better if the commits were squashed before merging.

Done.

tovl reviewed (approved):
LGTM :+1:

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5f6f9b5db6d20a0b12a691d24eddb3bb5550bda5
